Client/Frontend/Menu/MenuConfiguration.swift

```diff
@@ -4,146 +4,206 @@
 
 import Foundation
 
-enum MenuLocation {
-    case Browser
-    case HomePanels
-    case TabTray
+protocol MenuConfiguration {
+
+    var menuItems: [MenuItem] { get }
+    var menuToolbarItems: [MenuToolbarItem]? { get }
+    var numberOfItemsInRow: Int { get }
+
+    init(appState: AppState)
+    func toolbarColour() -> UIColor
+    func toolbarTintColor() -> UIColor
+    func menuBackgroundColor() -> UIColor
+    func menuTintColor() -> UIColor
+    func menuFont() -> UIFont
+    func menuIcon() -> UIImage?
+    func shadowColor() -> UIColor
+    func selectedItemTintColor() -> UIColor
 }
 
-struct MenuConfiguration {
 
-    let menuItems: [MenuItem]
-    let menuToolbarItems: [MenuToolbarItem]?
-    let location: MenuLocation
-    let numberOfItemsInRow: Int
+protocol MenuActionDelegate: class {
+    func performMenuAction(action: MenuAction, withAppState appState: AppState)
+}
+
+enum MenuAction {
+    case OpenNewNormalTab
```

Comment:
What gets passed around might be an enum, but what gets baked in to the menu shouldn't be. At the least this should be a marker protocol, that each menu's client casts to a specialst enum.

```swift
protocol MenuAction {}
enum TabMenuAction: MenuAction {}
```